ext/js/comm/shared-worker-bridge.js

```diff
@@ -0,0 +1,85 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import {createApiMap, invokeApiMapHandler} from '../core/api-map.js';
+import {log} from '../core/log.js';
+
+/**
+ * This serves as a bridge between the application and the backend.
+ * It is designed to have extremely short lifetime on the application side,
+ * as otherwise it will stay alive across extension updates (which only restart
+ * the backend) which can lead to extremely difficult to debug situations where
+ * the bridge is running an old version of the code.
+ *
+ * All it does is broker a handshake between the application and the backend,
+ * where they establish a connection between each other with a MessageChannel.
+ *
+ * # On backend startup
+ *  backend
+ *    ↓↓<"registerBackendPort" via SharedWorker.port.postMessage>↓↓
+ *  bridge: store the port in state
+ *
+ * # On application startup
+ *  application: create a new MessageChannel, bind event listeners to one of the ports, and send the other port to the bridge
+ *    ↓↓<"connectToBackend1" via SharedWorker.port.postMessage>↓↓
+ *  bridge
+ *    ↓↓<"connectToBackend2" via MessageChannel.port.postMessage which is stored in state from backend startup phase>↓↓
+ *  backend: bind event listeners to the other port
+ */
+export class SharedWorkerBridge {
+    constructor() {
+        /** @type {MessagePort?} */
+        this._backendPort = null;
+
+        /** @type {import('shared-worker').ApiMap} */
+        this._apiMap = createApiMap([
+            ['registerBackendPort', this._onRegisterBackendPort.bind(this)],
+            ['connectToBackend1', this._onConnectToBackend1.bind(this)],
+        ]);
+    }
+
+    /**
+     *
+     */
+    prepare() {
+        addEventListener('connect', (connectEvent) => {
+            const interlocutorPort = (/** @type {MessageEvent} */ (connectEvent)).ports[0];
+            interlocutorPort.addEventListener('message', (/** @type {MessageEvent<import('shared-worker').ApiMessageAny>} */ event) => {
+                const {action, params} = event.data;
+                return invokeApiMapHandler(this._apiMap, action, params, [interlocutorPort, event.ports], () => {});
+            });
+            interlocutorPort.start();
+        });
+    }
+
+    /** @type {import('shared-worker').ApiHandler<'registerBackendPort'>} */
+    _onRegisterBackendPort(_params, interlocutorPort, _ports) {
+        this._backendPort = interlocutorPort;
+    }
+
+    /** @type {import('shared-worker').ApiHandler<'connectToBackend1'>} */
+    _onConnectToBackend1(_params, _interlocutorPort, ports) {
+        if (this._backendPort !== null) {
+            this._backendPort.postMessage(void 0, [ports[0]]); // connectToBackend2
```

Comment:
I originally was doing that, but basically either you choose the fully typesafe API style of doing things like I did everywhere else, which requires a ton of extra code (like maybe 50~100 lines), but it seemed like overkill when there is only a single type of message which gets sent over this channel which has no (normal) parameters, so I decided to simplify it, since I can't imagine this ever getting more message types in the future (since the whole purpose of it is to establish a channel that you can use to send properly typed messages of any kind that you want). If you think it's too confusing I guess we can introduce the full API boilerplate, but we have a lot of different APIs due to this PR and I think introducing yet another one might be bad unless we figure out some better way to name them to make it clearer what's going on.